synapse/cores/storage.py

```diff
@@ -0,0 +1,827 @@
+        self.release()
+        time.sleep(0)
+        self.acquire()
+
+    def fire(self, name, **props):
+        '''
+        Pend an event to fire when the transaction next commits.
+        '''
+        self.events.append((name, props))
+
+        if self.size is not None and len(self.events) >= self.size:
+            self.sync()
+            self.cedetime()
+            self.begin()
+
+    def sync(self):
+        '''
+        Loop commiting and syncing events until there are no more
+        events that need to fire.
+        '''
+        self.commit()
+
+        # odd thing during exit... we need to fire events
+        # ( possibly causing more xact uses ) until there are
+        # no more events left to fire.
+        while self.events:
+            self.begin()
+            self.fireall()
+            self.commit()
+
+    def __enter__(self):
+        self.refs += 1
+        if self.refs == 1 and not self.ready:
+            self._coreXactInit()
+            self.acquire()
+            self.begin()
+            self.ready = True
+
+        return self
+
+    def __exit__(self, exc, cls, tb):
+        # FIXME handle rollback on exc not None
+        self.refs -= 1
+        if self.refs > 0 or self.exiting:
+            return
+
+        self.exiting = True
+
+        self.sync()
+        self.release()
+        self._coreXactFini()
+        self.store._popCoreXact()
+
+class Storage(s_config.Config):
+    '''
+    Base class for Cortex storage layer backends.
+
+    This implements functionality which is neccesary for a storage layer to
+    operate, as well as providing stubs for the storage layer implementations
+    to override.  See the Synapse Documentation for more details.
+
+    Args:
+        link ((str, dict)): Link tufo containing information for creating the Storage object.  This may include path
+                            information, authentication information, etc.
+        **conf (dict):  Additional configible options for the storage layer.
+    '''
+    def __init__(self,
+                 link,
+                 **conf):
+        s_config.Config.__init__(self)
+        if conf:
+            self.setConfOpts(conf)
+
+        #############################################################
+        # buses to save/load *raw* save events
+        #############################################################
+        self.savebus = s_eventbus.EventBus()
+        self.loadbus = s_eventbus.EventBus()
+
+        self._link = link
+
+        # XXX Can we eliminate this prop normalization need?
+        # We do need to know how to do prop normalization/defs for some helpers.
+        self.getPropNorm = None
+        self.getPropDef = None
+        # We need to be able to register tufosBy helpers
+        # which are storage helpers
+        # XXX Alternatively, provided a way for the cortex
+        # XXX to call into the storage layer and ask it
+        # To provide the tufo helper funcs?
+        self.initTufosBy = None
```

Comment:
so we have several "ways to lift tufos which are storage specific" - a easy solution was just to provide a per storage layer dictionary and have the Cortex register helpers through it.